### DISCUSSION TITLE
optimise training.iob_utils.offsets_to_biluo_tags method

### DISCUSSION THREAD

**Discussion by deepcode-debug** [explosion/spaCy#11441] [Help: Other Questions]
## How to reproduce the behaviour

run space.training.Example.from_dict 

the method offsets_to_biluo_tags in training.iob_utils is not optimised, it is taking quite significant time while running Example.from_dict 

Here is a profile snapshot

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    14601   24.677    0.002   90.797    0.006 {built-in method from_dict}
112839144   19.502    0.000   33.350    0.000 doc.pyx:484(__iter__)
        2   15.860    7.930   15.860    7.930 {built-in method _pickle.load}
112702047   13.855    0.000   13.855    0.000 token.pxd:21(cinit)
    15276    9.896    0.001   22.981    0.002 doc.pyx:1327(to_dict)
 18106173    7.784    0.000   11.133    0.000 vocab.pyx:142(get)
        1    7.507    7.507   41.008   41.008 {built-in method _pickle.dump}
   537703    7.312    0.000    7.312    0.000 decoder.py:343(raw_decode)
    11180    6.963    0.001   24.228    0.002 iob_utils.py:63(offsets_to_biluo_tags)
    31407    6.523    0.000   15.889    0.001 doc.pyx:180(__init__)
        1    5.164    5.164  181.599  181.599 Ned_process_deep.py:1(<module>)
  7968792    4.563    0.000    7.044    0.000 doc.pyx:694(genexpr)
    20025    4.461    0.000    7.149    0.000 __init__.py:100(get_all)
## Your Environment

* Operating System: Mac
* Python Version Used: 3.9
* spaCy Version Used: 3.3.1
* Environment Information: condo

**Comment by rmitsch** [comment]
Converting this to a discussion, as I don't think this is a concrete issue and a discussion might be a better format.

**Comment by deepcode-debug** [comment]
![image](https://user-images.githubusercontent.com/13822074/189070112-29dba1ab-7433-4d5a-9319-a4611c153071.png)

the problem is with the large loop in line 130: 
an alternative could be to initilialize all with "O" tokens and then assign only the entity values.

**Comment by adrianeboyd** [comment]
`Example.from_dict` is mainly intended as a transition method to convert the format used in v2 demo examples to the v3 `Doc`-based training data format `DocBin`/`.spacy`. I don't have enough context from this thread, but my guess is that you're concerned about the speed because you're using `Example.from_dict` within a training loop?

I would recommend converting your data once, storing it in a `DocBin`, saving it as a `.spacy` file to use with `spacy train` (or you can also read from it with `DocBin` or `Corpus` in custom training scripts, but we strongly recommend using a config with `spacy train`).

An example of a conversion script for NER data is here:

- data: https://github.com/explosion/projects/blob/v3/pipelines/ner_demo/assets/train.json
- conversion script: https://github.com/explosion/projects/blob/v3/pipelines/ner_demo/scripts/convert.py

A PR that improves the speed would always be welcome!
